tunnel/tunnel.go

```diff
@@ -79,165 +86,45 @@ func StartReverseTunnel(tunnelConfig *Config, wg *sync.WaitGroup, semaphore *Sem
+	}
+	fmt.Println(outputString)
 	var listener net.Listener
-
-	sshPort := tunnelConfig.TunnelEndpoint.SSHPort
-	// FIXME:  This should be a LUT
-	remoteEndpointPort := "3000"
-
-	if tunnelConfig.EndpointType == "https" {
-		remoteEndpointPort = "3001"
-	}
-
-	var jumpServerEndpoint = Endpoint{
-		Host: tunnelConfig.ConnectionEndpoint.Hostname(),
-		Port: tunnelConfig.ConnectionEndpoint.Port(),
-	}
-
-	// remote SSH server
-	var serverEndpoint = Endpoint{
-		Host: tunnelConfig.TunnelEndpoint.IPAddress,
-		Port: sshPort,
-	}
-
-	// remote forwarding port (on remote SSH server network)
-	var remoteEndpoint = Endpoint{
-		Host: "localhost", // localhost here is the remote SSHD daemon container
-		Port: remoteEndpointPort,
-	}
-
-	privateKey, err := readPrivateKeyFile(tunnelConfig.PrivateKeyPath)
-	if err != nil {
-		return listener, err
-	}
-	hostKeyCallBack := dnsHostKeyCallback
-	if tunnelConfig.ConnectionEndpoint.Hostname() != "api.holepunch.io" {
-		fmt.Println("Ignoring hostkey")
-		hostKeyCallBack = ssh.InsecureIgnoreHostKey()
-	}
-	sshJumpConfig := &ssh.ClientConfig{
-		User: "punch",
-		Auth: []ssh.AuthMethod{
-			ssh.Password(""),
-		},
-		HostKeyCallback: hostKeyCallBack,
-		Timeout:         0,
-	}
-
-	log.Debugf("Dial into Jump Server %s", jumpServerEndpoint.String())
-	jumpConn, err := ssh.Dial("tcp", jumpServerEndpoint.String(), sshJumpConfig)
-
+	listener, err = sClient.Listen("tcp", remoteEndpoint.String())
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error contacting the Holepunch Server.")
-		log.Debugf("%s", err)
-		return listener, err
+		fmt.Printf("%s", err.Error())
 	}
 
-	tunnelStartingSpinner(semaphore, &tunnelCreating)
-	exponentialBackoff := backoff.NewExponentialBackOff()
-
-	// Connect to SSH remote server using serverEndpoint
-	var serverConn net.Conn
 	for {
-		serverConn, err = jumpConn.Dial("tcp", serverEndpoint.String())
-		log.Debugf("Dial into SSHD Container %s", serverEndpoint.String())
-		if err == nil {
-			tunnelCreating.state = tunnelDone
-			break
+		// Open a (local) connection to localEndpoint whose content will be forwarded so serverEndpoint
+		log.Debugf("Dial to local %s", localEndpoint.String())
+		_, errInitialRemoteConnect := net.Dial("tcp", localEndpoint.String())
+		if errInitialRemoteConnect == nil {
+			client, errClient := listener.Accept()
+			remote, errRemote := net.Dial("tcp", localEndpoint.String())
+			if errRemote == nil && errClient == nil && client != nil && remote != nil {
+				// start goroutine
+				go handleClient(client, remote)
+			}
+		} else {
+			log.Debugf("Err %s", errInitialRemoteConnect.Error())
+			listener.Close()
+			listener = nil // you can't close the underlying file descriptor on the connection
+			// so you need to let the listener be GC'ed by replacing it with a new object
+			log.Debugf("No local listener")
+			time.Sleep(1000 * time.Millisecond)
+			log.Debugf("Trying again")
+			listener, _ = sClient.Listen("tcp", remoteEndpoint.String())
```

Comment:
I'm not sure we should ignore this error.  I can't really imagine something recoverable here, but I can envision a failure putting the client into a state where it panics or gives bad feedback.